File: ios/Classes/Bridge/AdSizeMethodHandler.swift
//
//  AdSizeMethodHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions
import Flutter

class AdSizeMethodHandler: CASChannel {
    init(with registrar: FlutterPluginRegistrar) {
        super.init(with: registrar, on: "ad_size")
    }

    override func onMethodCall(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "getInlineBanner": getInlineBanner(call, result)
        case "getAdaptiveBanner": getAdaptiveBanner(call, result)
        case "getAdaptiveBannerInScreen": getAdaptiveBannerInScreen(call, result)
        case "getSmartBanner": getSmartBanner(call, result)
        default: super.onMethodCall(call, result)
        }
    }

    private func getInlineBanner(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturnResult<Int, Int>("width", "maxHeight", result) { (width: Int, maxHeight: Int) in
            CASSize.getInlineBanner(width: CGFloat(width), maxHeight: CGFloat(maxHeight)).toMap()
        }
    }

    private func getAdaptiveBanner(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturnResult<Int>("maxWidthDp", result) { (maxWidthDp: Int) in
            CASSize.getAdaptiveBanner(forMaxWidth: CGFloat(maxWidthDp)).toMap()
        }
    }

    private func getAdaptiveBannerInScreen(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        result(CASSize.getAdaptiveBanner(inWindow: UIApplication.shared.delegate!.window!!).toMap())
    }

    private func getSmartBanner(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        result(CASSize.getSmartBanner().toMap())
    }
}

extension CASSize {
    func toMap() -> [String: Int] {
        let mode: Int
        if isAdaptive { mode = 2 }
        else if isInline { mode = 3 }
        else { mode = 0 }
        return [
            "width": Int(width),
            "height": Int(height),
            "mode": mode,
        ]
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bridge/AdSizeMethodHandler.kt
package com.cleveradssolutions.plugin.flutter.bridge

import com.cleveradssolutions.plugin.flutter.CASFlutterContext
import com.cleveradssolutions.plugin.flutter.bridge.base.MethodHandler
import com.cleveradssolutions.plugin.flutter.util.getArgAndReturnResult
import com.cleversolutions.ads.AdSize
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val CHANNEL_NAME = "cleveradssolutions/ad_size"

class AdSizeMethodHandler(
    binding: FlutterPluginBinding,
    private val contextService: CASFlutterContext
) : MethodHandler(binding, CHANNEL_NAME) {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getInlineBanner" -> getInlineBanner(call, result)
            "getAdaptiveBanner" -> getAdaptiveBanner(call, result)
            "getAdaptiveBannerInScreen" -> getAdaptiveBannerInScreen(result)
            "getSmartBanner" -> getSmartBanner(result)
            else -> super.onMethodCall(call, result)
        }
    }

    private fun getInlineBanner(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturnResult<Int, Int>("width", "maxHeight", result) { width, maxHeight ->
            AdSize.getInlineBanner(width, maxHeight).toMap()
        }
    }

    private fun getAdaptiveBanner(call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturnResult<Int>("maxWidthDp", result) { maxWidthDp ->
            AdSize.getAdaptiveBanner(contextService.getContext(), maxWidthDp).toMap()
        }
    }

    private fun getAdaptiveBannerInScreen(result: MethodChannel.Result) {
        result.success(
            AdSize.getAdaptiveBannerInScreen(contextService.getContext()).toMap()
        )
    }

    private fun getSmartBanner(result: MethodChannel.Result) {
        result.success(
            AdSize.getSmartBanner(contextService.getContext()).toMap()
        )
    }

    private fun AdSize.toMap(): Map<String, Int> {
        return mapOf(
            "width" to width,
            "height" to height,
            "mode" to when {
                isAdaptive -> 2
                isInline -> 3
                else -> 0
            }
        )
    }

}